### ISSUE TITLE
core dump in native code with LSTM and rc3.9

### ISSUE THREAD

**Issue by fac2003** [deeplearning4j/deeplearning4j#1556]
I am experiencing reproducible core dumps when running a LSTM training similar to the GravesLSTMCharModellingExample sample. I am happy to share the code and the data if you need it to track down the issue. 
```

#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000012cf8b6e9, pid=85398, tid=5891
#
# JRE version: Java(TM) SE Runtime Environment (8.0_40-b27) (build 1.8.0_40-b27)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.40-b25 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libnd4j.dylib+0x236e9]  _ZN5shape3TAD4initEPiS1_i+0x249
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/fac2003/IdeaProjects/git/dl4c/hs_err_pid85398.log
#
```

Here's the relevant content of hs_err:

```

Stack: [0x000070000011a000,0x000070000021a000],  sp=0x0000700000218400,  free space=1017k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libnd4j.dylib+0x236e9]  shape::TAD::init(int*, int*, int)+0x249
C  [libnd4j.dylib+0x279ef]  functions::reduce::ReduceFunction<float>::exec(float*, int*, float*, float*, int*, int*, int)+0xff
C  [libnd4j.dylib+0x63aa]  NativeOps::execReduceFloat(long long*, int, long long, long long, long long, long long, long long, long long, int)+0x5a
C  [libjnind4j.dylib+0xe6e9]  Java_org_nd4j_nativeblas_NativeOps_execReduceFloat___3JIJJJJJJI+0xb9
J 1461  org.nd4j.nativeblas.NativeOps.execReduceFloat([JIJJJJJJI)V (0 bytes) @ 0x000000010d7dfa31 [0x000000010d7df940+0xf1]
J 1986 C2 org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(Lorg/nd4j/linalg/api/ops/Accumulation;[I)Lorg/nd4j/linalg/api/ndarray/INDArray; (1441 bytes) @ 0x000000010d9f5938 [0x000000010d9f4ec0+0xa78]

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
J 1461  org.nd4j.nativeblas.NativeOps.execReduceFloat([JIJJJJJJI)V (0 bytes) @ 0x000000010d7df9b7 [0x000000010d7df940+0x77]
J 1986 C2 org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(Lorg/nd4j/linalg/api/ops/Accumulation;[I)Lorg/nd4j/linalg/api/ndarray/INDArray; (1441 bytes) @ 0x000000010d9f5938 [0x000000010d9f4ec0+0xa78]
J 1985 C2 org.nd4j.linalg.api.ndarray.BaseNDArray.sum([I)Lorg/nd4j/linalg/api/ndarray/INDArray; (18 bytes) @ 0x000000010d9d9dfc [0x000000010d9d9b20+0x2dc]
j  org.deeplearning4j.nn.layers.recurrent.LSTMHelpers.backpropGradientHelper(Lorg/deeplearning4j/nn/conf/NeuralNetConfiguration;Lorg/nd4j/linalg/api/ndarray/INDArray;Lorg/nd4j/linalg/api/ndarray/INDArray;Lorg/nd4j/linalg/api/ndarray/INDArray;Lorg/nd4j/linalg/api/ndarray/INDArray;ZILorg/deeplearning4j/nn/layers/recurrent/FwdPassReturn;ZLjava/lang/String;Ljava/lang/String;Ljava/lang/String;)Lorg/deeplearning4j/berkeley/Pair;+1567
j  org.deeplearning4j.nn.layers.recurrent.GravesLSTM.backpropGradientHelper(Lorg/nd4j/linalg/api/ndarray/INDArray;ZI)Lorg/deeplearning4j/berkeley/Pair;+127
j  org.deeplearning4j.nn.layers.recurrent.GravesLSTM.backpropGradient(Lorg/nd4j/linalg/api/ndarray/INDArray;)Lorg/deeplearning4j/berkeley/Pair;+4
j  org.deeplearning4j.nn.multilayer.MultiLayerNetwork.calcBackpropGradients(Lorg/nd4j/linalg/api/ndarray/INDArray;Z)Lorg/deeplearning4j/berkeley/Pair;+330
j  org.deeplearning4j.nn.multilayer.MultiLayerNetwork.backprop()V+3
j  org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore()V+38
j  org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore()Lorg/deeplearning4j/berkeley/Pair;+12
j  org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize()Z+14
j  org.deeplearning4j.optimize.Solver.optimize()V+19

```

**Comment by raver119** [comment]
So, that's Mac OS X, nd4-native backend, right?

**Comment by fac2003** [comment]
Correct.

**Comment by raver119** [comment]
We'll need to see your source code and pom.xml. Please use http://gist.github.com to upload them and post links to them here

**Comment by fac2003** [comment]
I've shared the small repo on BitBucket if you want to have a look (with raver119, I can add more users). The class is TrainLSTMFromBits. You can run it with one arg: data/latex/all.tex, it fails for me usually after iteration 40, but I have seen failures as far as iteration 71

**Comment by raver119** [comment]
Can i have link to that repo?

**Comment by fac2003** [comment]
https://bitbucket.org/campagnelaboratory/dl4c

**Comment by raver119** [comment]
"after iteration 40" sounds suspicious... I'll take a look later, thanks.

**Comment by fac2003** [comment]
I ran this for hours with rc3.8 without a glitch.

**Comment by fac2003** [comment]
In case anybody cares, I am running this on: 
java version "1.8.0_65"
Java(TM) SE Runtime Environment (build 1.8.0_65-b17)
Java HotSpot(TM) 64-Bit Server VM (build 25.65-b01, mixed mode)

**Comment by fac2003** [comment]
I was able to reproduce this with the GravesLSTMCharModelingExample with the following changes to hyperparameters:
		int exampleLength = 300;			
and 
.learningRate(0.01)	

I then get: 

```
o.d.o.l.ScoreIterationListener - Score at iteration 414 is 149.3247007444239
o.d.o.l.ScoreIterationListener - Score at iteration 415 is 144.7411355886625
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000012f1a56e9, pid=63580, tid=5891
#
# JRE version: Java(TM) SE Runtime Environment (8.0_40-b27) (build 1.8.0_40-b27)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.40-b25 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libnd4j.dylib+0x236e9]  shape::TAD::init(int*, int*, int)+0x249
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/fac2003/IdeaProjects/git/dl4j-0.4-examples/hs_err_pid63580.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

Process finished with exit code 134
```

**Comment by AlexDBlack** [comment]
@fac2003 thanks, I'll be looking into this today.

**Comment by AlexDBlack** [comment]
Ok, so with those hyperparameters on the LSTM example: No issue for me (Windows 8.1); I ran through the full example.
So it may be platform specific.

@nyghtowl you are still on OSX, right? Could you try running this, see if it occurs for you?

**Comment by fac2003** [comment]
I tried this on two different macs (a laptop, a mac pro desktop), both running 10.11.4. Same errors on these two machines.

**Comment by AlexDBlack** [comment]
One question though: does it always occur at the exact same iteration? Or does it occur at different times during training?

**Comment by fac2003** [comment]
Different steps with the same random seed, but I use multithreading, which could impact the iteration by timing differences.

**Comment by nyghtowl** [comment]
I also got core dump on my Mac for GravesLSTM but Linux was fine. @eraly can you run this on your machine tomorrow and confirm if you are getting the same behavior?

**Comment by agibsonccc** [comment]
Could we try to narrow down a cause as well? I'd like to get a test case we
can put in to nd4j.

**Comment by AlexDBlack** [comment]
J 1461  org.nd4j.nativeblas.NativeOps.execReduceFloat([JIJJJJJJI)V (0 bytes) @ 0x000000010d7df9b7 [0x000000010d7df940+0x77]
J 1986 C2 org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(Lorg/nd4j/linalg/api/ops/Accumulation;[I)Lorg/nd4j/linalg/api/ndarray/INDArray; (1441 bytes) @ 0x000000010d9f5938 [0x000000010d9f4ec0+0xa78]
J 1985 C2 org.nd4j.linalg.api.ndarray.BaseNDArray.sum([I)Lorg/nd4j/linalg/api/ndarray/INDArray; (18 bytes) @ 0x000000010d9d9dfc [0x000000010d9d9b20+0x2dc]

So, it's a sum(0) - that's the only sum along dimensions that are done in LSTMs. They are always on f order, too.

Running the full nd4j test suite on a mac is probably a good start, in case we already have something that picks this up.

**Comment by fcecconi** [comment]
I have the same problem in linux:

Stack: [0x00007fe413be9000,0x00007fe413cea000],  sp=0x00007fe413ce7c90,  free space=1019k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libnd4j.so+0x65353]  functions::reduce::ReduceFunction<float>::exec(float*, int*, float*, float*, int*, int*, int) [clone ._omp_fn.56]+0x233

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
J 3802  org.nd4j.nativeblas.NativeOps.execReduceFloat([JIJJJJJJI)V (0 bytes) @ 0x00007fe3fd9d9c37 [0x00007fe3fd9d9bc0+0x77]
J 4352 C2 org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(Lorg/nd4j/linalg/api/ops/Accumulation;[I)Lorg/nd4j/linalg/api/ndarray/INDArray; (1441 bytes) @ 0x00007fe3fd9562c4 [0x00007fe3fd955740+0xb84]
J 4251 C2 org.nd4j.linalg.api.ndarray.BaseNDArray.sum([I)Lorg/nd4j/linalg/api/ndarray/INDArray; (18 bytes) @ 0x00007fe3fd89c6bc [0x00007fe3fd89c3e0+0x2dc]
J 6970 C2 org.deeplearning4j.nn.layers.recurrent.LSTMHelpers.backpropGradientHelper(Lorg/deeplearning4j/nn/conf/NeuralNetConfiguration;Lorg/nd4j/linalg/api/ndarray/INDArray;Lorg/nd4j/linalg/api/ndarray/INDArray;Lorg/nd4j/linalg/api/ndarray/INDArray;Lorg/nd4j/linalg/api/ndarray/INDArray;ZILorg/deeplearning4j/nn/layers/recurrent/FwdPassReturn;ZLjava/lang/String;Ljava/lang/String;Ljava/lang/String;)Lorg/deeplearning4j/berkeley/Pair; (1986 bytes) @ 0x00007fe3fe2a1f34 [0x00007fe3fe29e680+0x38b4]
J 6896 C1 org.deeplearning4j.nn.layers.recurrent.GravesLSTM.backpropGradientHelper(Lorg/nd4j/linalg/api/ndarray/INDArray;ZI)Lorg/deeplearning4j/berkeley/Pair; (131 bytes) @ 0x00007fe3fd383cc4 [0x00007fe3fd383060+0xc64]
J 6895 C1 org.deeplearning4j.nn.layers.recurrent.GravesLSTM.backpropGradient(Lorg/nd4j/linalg/api/ndarray/INDArray;)Lorg/deeplearning4j/berkeley/Pair; (8 bytes) @ 0x00007fe3fd481fcc [0x00007fe3fd481f40+0x8c]
J 6890 C1 org.deeplearning4j.nn.multilayer.MultiLayerNetwork.calcBackpropGradients(Lorg/nd4j/linalg/api/ndarray/INDArray;Z)Lorg/deeplearning4j/berkeley/Pair; (648 bytes) @ 0x00007fe3fe25b7bc [0x00007fe3fe258a20+0x2d9c]
J 7367 C1 org.deeplearning4j.nn.multilayer.MultiLayerNetwork.backprop()V (46 bytes) @ 0x00007fe3fdc1e064 [0x00007fe3fdc1dfc0+0xa4]
j  org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore()V+38
J 7693 C1 org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore()Lorg/deeplearning4j/berkeley/Pair; (67 bytes) @ 0x00007fe3fd1d17bc [0x00007fe3fd1d16a0+0x11c]
J 7692 C1 org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize()Z (147 bytes) @ 0x00007fe3fde1df54 [0x00007fe3fde1de40+0x114]
J 7691 C1 org.deeplearning4j.optimize.Solver.optimize()V (26 bytes) @ 0x00007fe3fdd12ccc [0x00007fe3fdd12b20+0x1ac]
j  org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(Lorg/nd4j/linalg/dataset/api/iterator/DataSetIterator;)V+272

**Comment by saudet** [comment]
Pull https://github.com/deeplearning4j/libnd4j/pull/207 should fix this.

**Comment by eraly** [comment]
@fac2003 can you try this again when we release 3.10 is out? Or if you are impatient you could always build from source? Thanks!

**Comment by fac2003** [comment]
I am trying this with 3.10 and it seems to hold up so far (one hour into training with 8 threads). I'll write only if I get a failure.
